<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
  <head>
    <title>SpecID</title>
    <link href="../default.css" rel="stylesheet" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  </head>
  <body>
    <div id="logo">
        <a href=../index.html><img src=../images/mspire_logo.png border="0" alt="mspire - mass spectrometry proteomics in ruby"/></a>
  </div>
    <div id="header">
      <h1>SpecID</h1>
    </div>

    <div id="headerbar" class="bar">
        <!-- <span class="left">Location: <a href="../index.html" title="mspire">mspire</a> / <a href="./index.html" title="SpecID">SpecID</a> / <a href="./spec_id.html" title="SpecID">SpecID</a></span> -->
      <!--<span class="right">Language: <a href="./spec_id.html" title="en">en</a></span>-->
      <div style="clear:both"></div>
    </div>

    <div id="menu">
        <div class="webgen-menu-vert "><ul><li ><a href="../index.html" title="Home">Home</a></li><li class="webgen-menu-submenu"><a href="../install/index.html" title="Install">Install</a></li><li class="webgen-menu-submenu"><a href="../ms/index.html" title="MS">MS</a></li><li class="webgen-menu-submenu webgen-menu-submenu-inhierarchy"><a href="./index.html" title="SpecID">SpecID</a><ul><li ><a href="./srf.html" title="SRF">SRF</a></li><li class="webgen-menu-item-selected"><a href="./spec_id.html" title="SpecID">SpecID</a></li><li class="webgen-menu-submenu"><a href="fir/index.html" title="FIR">FIR</a></li></ul></li><li class="webgen-menu-submenu"><a href="../tutorial/index.html" title="Tutorials">Tutorials</a></li></ul></div>
        <a style="font-size:130%; padding-left:20px" href="rdoc/index.html">rdoc</a>
        <br/>
        <a style="font-size:85%; padding-left:20px" href="http://rubyforge.org/projects/mspire/">Rubyforge Project Page</a>
        <br/>
        <br/>
        <br/>
        <br/>
        <a href="http://www.ruby-lang.org/en/"><img style="padding-left:20px" src=../images/ruby_logo_small.png border="0" alt="Ruby Programming Language"/></a>
    </div>

    <div id="body">
      <h1>Object Model</h1>

<p>SpecID is a mixin (or can be thought of as an interface) for objects containing peptide and/or protein identifications. </p>

<p><img src="../images/spec_id_object_model.png" alt="SpecID object model" title="SpecID Object model"/></p>

<h1>Examples</h1>

<h2>Reading spectral ID files:</h2>

<p>mspire can read several different spectral ID formats.  Each is farmed out to
a format specific object that mixes in the SpecID module and whose
identifications mixin the SpecID::Pep or SpecID::Prot module.</p>

<pre><code>require "spec_id" 

specid = SpecID.new(["file1.srf", "file2.srf"]) # array of srf files
specid = SpecID.new("file1.srf")            # single srf file
specid = SpecID.new("bioworks.xml")         # Bioworks multiconcesus output
specid = SpecID.new("peptide_prophet.xml")  # Peptide Prophet
specid = SpecID.new("file.sqt")             # a sqt file
specid = SpecID.new("percolator.sqt")       # a sqt file after percolator
specid = SpecID.new("file.sqg")             # a sqt group file

specid.peps          # =&gt; an array of peptide identifications
specid.prots         # =&gt; an array of protein identifications
specid.hi_prob_best  # =&gt; boolean: high probability best?
                     # peptide prophet      == true
                     # percolator (q-value) == true
                     # bioworks             == false

specid.filename      # =&gt; the file it was derived from
</code></pre>

<p>Each of the parsers for these formats ensures that each peptide ID links up to
its associated proteins and each protein ID links to associated peptide ID's
<em>with no redundancy</em>.  The SpecID module means that you can get at the ID's
much faster than having to traverse the proper object heirarchy that is
built for each object type.  </p>

<p>A simple example will demonstrate how much simpler it is to use the SpecID
module rather than traversing the object heirarchy (which is available
if you need it):</p>

<pre><code># using the SQT object heirarchy

sqg = SQTGroup.new("file.sqg")
sqg_object.sqts.each do |sqt|
  sqt.spectra.each do |spectrum|
    spectrum.each do |match|
      match.xcorr                # &lt;- do something with the match
      match.loci.each do |locus|
        locus.reference           # &lt;- do something with each protein ID
      end
    end
  end
end

# using the SpecID module

specid = SpecID.new("file.sqg")
specid.peps.each do |match|     # &lt;- a pep is really a SQT::Match!
  match.xcorr
end

specid.prots.each do |locus|   # &lt;- a prot is really a SQT::Locus!
  locus.reference
end
</code></pre>

<p>This kind of shared interface is especially handy when you are dealing with
different file formats.</p>

<h3>SpecID::Pep</h3>

<p>A SpecID::Pep responds to basic information about the identification:</p>

<pre><code>specid.peps.each do |pep|
  pep.prots         # SpecID::Prot proteins associated with the peptide
  pep.sequence      # a sequence showing leading and tailing
                    # amino acids, e.g.:  K.AASEQENCE.A  ('-' for none)
  pep.aaseq         # the bare amino acid sequence, e.g.:  AASEQENCE 
  pep.charge        # charge of the hit peptide
  pep.probability   # probability of the hit being accurate
end

specid.peps.sort    # sorted by aaseq (&lt;=&gt; defined in terms of aaseq)

# sort from worst to best hit:
sorted = specid.peps.sort_by {|pep| pep.probability }
unless specid.hi_prob_best
  sorted.reverse!
end

SpecID::Pep.protein_groups_by_sequence(['AASEQ', 'PEP', 'TIDE'], file.fasta)
</code></pre>

<h3>SpecID::Prot</h3>

<p>A SpecID::Prot object responds to basic information about the protein information:</p>

<pre><code>specid.prots.each do |prot|
  prot.reference    # unique identifier (the fasta header until first space)
  prot.probability 
  prot.peps         # the peptide hits
end
</code></pre>

<h3>Bioworks files (srf and Bioworks xml output)</h3>

<pre><code>specid = SpecID.new(["file1.srf", "file2.srf", "file3.srf"])
specid.class         # =&gt; SRFGroup
specid.is_a? SpecID  # =&gt; true
</code></pre>

<h3>SQT files</h3>

<h3>PeptideProphet (pepXML)</h3>
    </div>

    <div id="footer" class="bar">
      Last update: <i>Monday, August 18 10:50:22 MDT 2008</i>
    </div>
  </body>
</html>
